### DISCUSSION TITLE
[Windows] PE overlay is not being compressed

### DISCUSSION THREAD

**Discussion by alonbl** [upx/upx#725] [General]
Hello UPX developers,

I am trying to pack EXE produced by PyInstaller using UPX.

PyInstaller is placing the payload as PE overlay (PE additional/extra data), hence the actual python payload is packaged within the overlay.

I expected all sections of the original EXE to be compress, however, as far as I can see UPX packs and compress all the PE sections, but leaves the overlay as-is.

I do see that UPX has the following options:
```
Overlay options:
  --overlay=copy      copy any extra data attached to the file [default]
  --overlay=strip     strip any extra data attached to the file [DANGEROUS]
  --overlay=skip      don't compress a file with an overlay
```

I now understand that "copy" means transparent copy, unlike any other section.

I would really like to have the overlay compress somehow :)

I would like to ask:

1. Is there any issue in compressing the overlay, I understand that the overlay is just a blob that resides after the PE size, have you find an issue of decompressing it before passing control to the entry point?

2. Was there any other reason that compressing the overlay cannot be done? I mean why only copy? why not compress?

3. If it can be done, can you please provide some hints to relevant code where this may be implemented? I may look at implementing this if I understand some of the past context, I could find:

   1. `PackExe::pack()` which copies the overlay and may also compress it.
   2. `PackW64PeAmd64::buildLoader()` which may have `addLoader()` to decompress it.

Thank you,
Alon Bar-Lev

As a reference, attached an example of such executable: [upx-test.zip](https://github.com/upx/upx/files/13165396/upx-test.zip)

**Comment by ml1050** [comment]
Hi,

The name "overlay data" means that the data is not loaded by the OS into the memory when it starts the program. So it's the responsibility of the application to load/decompress the data later by itself. UPX can not help here.

**Reply by alonbl** [reply]
Thank you for the quick reply!

If I understand what you say is that PyInstaller is opening the PE file by itself, parsing the header and seek to the overlay offset?

I will look into this in PyInstaller code.

**Reply by alonbl** [reply]
Yes, I found it in PyInstaller[1], thank you so much for the hint!

[1] https://github.com/pyinstaller/pyinstaller/blob/develop/bootloader/src/pyi_archive.c
